autonat.go

```diff
@@ -0,0 +1,142 @@
+package autonat
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	host "github.com/libp2p/go-libp2p-host"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type NATStatus int
+
+const (
+	NATStatusUnknown NATStatus = iota
+	NATStatusPublic
+	NATStatusPrivate
+)
+
+type AutoNAT interface {
+	Status() NATStatus
+	PublicAddr() (ma.Multiaddr, error)
+}
+
+type AutoNATState struct {
```

Comment:
the naming here is a bit weird. Its not clear from reading that `AutoNATState` is an implementation of `AutoNAT`